src/Compilers/CSharp/Portable/Symbols/FileIdentifier.cs

```diff
@@ -9,33 +9,91 @@
 
 namespace Microsoft.CodeAnalysis.CSharp;
 
-internal readonly struct FileIdentifier
+internal class FileIdentifier
 {
     private static readonly Encoding s_encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
 
-    public string? EncoderFallbackErrorMessage { get; init; }
-    public ImmutableArray<byte> FilePathChecksumOpt { get; init; }
-    public string DisplayFilePath { get; init; }
+    private string? _filePath;
 
-    public static FileIdentifier Create(SyntaxTree tree)
-        => Create(tree.FilePath);
+    private string? _encoderFallbackErrorMessage;
+    private string? _displayFilePath;
+    private ImmutableArray<byte> _filePathChecksumOpt;
 
-    public static FileIdentifier Create(string filePath)
+    private FileIdentifier(string filePath)
+    {
+        _filePath = filePath;
+    }
+
+    private FileIdentifier(ImmutableArray<byte> filePathChecksumOpt, string displayFilePath)
+    {
+        _filePathChecksumOpt = filePathChecksumOpt;
+        _displayFilePath = displayFilePath;
+    }
+
+    private void EnsureInitialized()
     {
-        string? encoderFallbackErrorMessage = null;
-        ImmutableArray<byte> hash = default;
-        try
+        var filePath = _filePath;
+        if (filePath is not null)
         {
-            var encodedFilePath = s_encoding.GetBytes(filePath);
-            using var hashAlgorithm = SourceHashAlgorithms.CreateDefaultInstance();
-            hash = hashAlgorithm.ComputeHash(encodedFilePath).ToImmutableArray();
+            string? encoderFallbackErrorMessage = null;
+            ImmutableArray<byte> hash = default;
+            try
+            {
+                var encodedFilePath = s_encoding.GetBytes(filePath);
+                using var hashAlgorithm = SourceHashAlgorithms.CreateDefaultInstance();
+                hash = hashAlgorithm.ComputeHash(encodedFilePath).ToImmutableArray();
+            }
+            catch (EncoderFallbackException ex)
+            {
+                encoderFallbackErrorMessage = ex.Message;
+            }
+
+            var displayFilePath = GeneratedNames.GetDisplayFilePath(filePath);
+
+            _encoderFallbackErrorMessage = encoderFallbackErrorMessage;
+            _filePathChecksumOpt = hash;
```

Comment:
>_filePathChecksumOpt = hash [](http://example.com/codeflow?start=12&length=27)

Assignment to structures is not atomic, therefore, not safe in a multi-threaded code. Consider using InterlockedInitialize.

Reply:
Also, consider doing the InterlockedInitialize right before the ```_filePath = null``` assignment to make sure the assignment isn't reordered.

Reply:
Using InterlockedInitialize now